flink-ml-lib/src/test/java/org/apache/flink/ml/common/datastream/AllReduceUtilsTest.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.datastream;
+
+import org.apache.flink.api.common.functions.FlatMapFunction;
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.NumberSequenceIterator;
+
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+import static org.junit.Assert.assertEquals;
+
+/** Tests the {@link org.apache.flink.ml.common.datastream.AllReduceUtils}. */
+@RunWith(Enclosed.class)
+public class AllReduceUtilsTest {
+
+    private static final int parallelism = 4;
+
+    /** Parameterized test for {@link org.apache.flink.ml.common.datastream.AllReduceUtils}. */
+    @RunWith(Parameterized.class)
+    public static class ParameterizedTest {
+        private static int numElements;
+
+        @Parameterized.Parameters
+        public static Collection<Object[]> params() {
+            return Arrays.asList(new Object[][] {{0}, {100}, {10000}, {100000}, {10000000}});
```

Comment:
Is there any semantic difference between these four parameters? If no, can we simplify the test (and its resource usage) by using just one config, e.g. `numElements = 100`?

Reply:
We need to test the following cases:
(1) when there are no chunks  (i.e., 0)
(2) when the data is not enough for one chunk (i.e., 100)
(3) when not every worker have one chunk to handle (i.e., 10000)
(4) when each worker needs to handle at least one chunk (i.e., 100000)

I have removed 10000000.

Reply:
The reason makes sense.

Could we add comment in the test explaining why we have these four `numElements` values? Even better, could we generate these `numElements` values  as a function of `parallelism` and `TRANSFER_BUFFER_SIZE`?

Reply:
Thanks for the comments. I have updated the PR. Please check out for AllReduceImplTest#L48

Reply:
Sounds good. Thanks for the update.